man/addtrans.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" (C) Copyright 2022 Marcin Owsiany <porridge@debian.org>,
.\"
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH addtrans 1 "November 11 2022"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
addtrans \- interactively add transactions to a ledger file
.SH SYNOPSIS
.B addtrans
.RI [ options ]
.SH DESCRIPTION
.B addtrans
is a graphical application for quickly adding new transactions to a ledger file.
.PP
A tutorial for using this program, with screenshots, is available at
.BR /usr/share/doc/ledgerhelpers/doc/addtrans.html
.PP
The program must be supplied with locations of the ledger and price database
files to work with.
The location of each file is determined independently, using the following
mechanisms, in this order.
The first mechanism which yields a result, wins.
.SH OPTIONS
.TP
.BR \-h ,
.BR \-\-help
Show help message and exit.
.TP
.B \-\-file FILE
Specify path to ledger file to work with.
.TP
.B \-\-price\-db PRICEDB
Specify path to ledger price database to work with.
.TP
.B \-\-debug
Turn on debugging output, may be useful for developers.
.SH ENVIRONMENT
The following environment variables are recognized by this program:
.TP
.BR LEDGER_FILE
Path to ledger file to work with.
.TP
.BR LEDGER_PRICE_DB
Path to ledger price database to work with.
.SH FILES
The config file for
.BR ledger (1),
namely file
.BR .ledgerrc
in user's home directory is scanned looking for the following options.
.TP
.B \-\-file FILE
Path to ledger file to work with.
.TP
.B \-\-price\-db PRICEDB
Path to ledger price database to work with.

.SH SEE ALSO
.BR ledger (1),
.BR cleartrans\-cli (1),
.BR sellstock\-cli (1),
.BR sorttrans\-cli (1),
.BR withdraw\-cli (1).
.br
A tutorial for using this program, with screenshots, is available at
.BR /usr/share/doc/ledgerhelpers/doc/addtrans.html


src/ledgerhelpers/programs/addtrans.py
#!/usr/bin/python3

import argparse
import datetime
import logging
import traceback

import gi
gi.require_version("Gtk", "3.0")
from gi.repository import GObject
from gi.repository import Gtk
from gi.repository import Pango

import ledgerhelpers as common
from ledgerhelpers import gui
from ledgerhelpers import journal
from ledgerhelpers.programs import common as common_programs
import ledgerhelpers.editabletransactionview as ed


ASYNC_LOAD_MESSAGE = "Loading completion data from your ledger..."
ASYNC_LOADING_ACCOUNTS_MESSAGE = (
    "Loading account and commodity data from your ledger..."
)


class AddTransWindow(Gtk.Window, gui.EscapeHandlingMixin):

    def __init__(self):
        Gtk.Window.__init__(self, title="Add transaction")
        self.set_border_width(12)

        grid = Gtk.Grid()
        grid.set_column_spacing(8)
        grid.set_row_spacing(12)
        self.add(grid)

        row = 0

        self.transholder = ed.EditableTransactionView()
        grid.attach(self.transholder, 0, row, 2, 1)

        row += 1

        self.transaction_view = gui.LedgerTransactionView()
        self.transaction_view.set_vexpand(True)
        self.transaction_view.get_accessible().set_name("Transaction preview")
        grid.attach(self.transaction_view, 0, row, 2, 1)

        row += 1

        self.status = Gtk.Label()
        self.status.set_line_wrap(True)
        self.status.set_line_wrap_mode(Pango.WrapMode.WORD_CHAR)
        self.status.set_hexpand(True)
        grid.attach(self.status, 0, row, 1, 1)

        button_box = Gtk.ButtonBox()
        button_box.set_layout(Gtk.ButtonBoxStyle.END)
        button_box.set_spacing(12)
        button_box.set_hexpand(False)
        self.close_button = Gtk.Button(stock=Gtk.STOCK_CLOSE)
        button_box.add(self.close_button)
        self.add_button = Gtk.Button(stock=Gtk.STOCK_ADD)
        button_box.add(self.add_button)
        grid.attach(button_box, 1, row, 1, 1)
        self.add_button.set_can_default(True)
        self.add_button.grab_default()


class AddTransApp(AddTransWindow, gui.EscapeHandlingMixin):

    logger = logging.getLogger("addtrans")
    internal_parsing = []

    def __init__(self, journal, preferences):
        AddTransWindow.__init__(self)
        self.journal = journal
        self.preferences = preferences
        self.successfully_loaded_accounts_and_commodities = False

        self.accounts = []
        self.commodities = dict()
        self.internal_parsing = []
        self.payees = []

        self.activate_escape_handling()

        self.close_button.connect("clicked",
                                  lambda _: self.emit('delete-event', None))
        self.add_button.connect("clicked",
                                lambda _: self.process_transaction())
        date = self.preferences.get("last_date", datetime.date.today())
        self.transholder.set_transaction_date(date)
        self.transholder.connect(
            "payee-changed",
            self.payee_changed
        )
        self.transholder.connect(
            "changed",
            self.update_transaction_view
        )

        self.add_button.set_sensitive(False)
        self.transholder.title_grab_focus()
        self.status.set_text(ASYNC_LOAD_MESSAGE)

        self.connect("delete-event", lambda _, _a: self.save_preferences())
        self.reload_completion_data()

    def reload_completion_data(self):
        gui.g_async(
            lambda: self.journal.internal_parsing(),
            lambda payees: self.internal_parsing_loaded(payees),
            self.journal_load_failed,
        )

    def internal_parsing_loaded(self, internal_parsing):
        self.internal_parsing = internal_parsing
        gui.g_async(
            lambda: self.journal.all_payees(),
            lambda payees: self.all_payees_loaded(payees),
            self.journal_load_failed,
        )

    def all_payees_loaded(self, payees):
        self.payees = payees
        self.transholder.set_payees_for_completion(self.payees)
        gui.g_async(
            lambda: self.journal.accounts_and_last_commodity_for_account(),
            lambda r: self.accounts_and_last_commodities_loaded(*r),
            self.journal_load_failed,
        )
        if self.status.get_text() == ASYNC_LOAD_MESSAGE:
            self.status.set_text(ASYNC_LOADING_ACCOUNTS_MESSAGE)

    def accounts_and_last_commodities_loaded(self, accounts, last_commos):
        self.accounts = accounts
        self.commodities = last_commos
        self.transholder.set_accounts_for_completion(self.accounts)
        self.transholder.set_default_commodity_getter(
            self.get_commodity_for_account
        )
        self.successfully_loaded_accounts_and_commodities = True
        if self.status.get_text() == ASYNC_LOADING_ACCOUNTS_MESSAGE:
            self.status.set_text("")

    def journal_load_failed(self, e):
        traceback.print_exception(e)
        gui.FatalError(
            "Add transaction loading failed",
            "An unexpected error took place:\n%s" % e,
        )
        self.emit('delete-event', None)

    def get_commodity_for_account(self, account_name):
        try:
            return self.commodities[account_name]
        except KeyError:
            pass

    def payee_changed(self, emitter=None):
        if emitter.postings_modified() and not emitter.postings_empty():
            return
        text = emitter.get_payee_text()
        self.try_autofill(emitter, text)

    def try_autofill(self, transaction_view, autofill_text):
        ts = journal.transactions_with_payee(
            autofill_text,
            self.internal_parsing,
            case_sensitive=False
        )
        if not ts:
            return
        return self.autofill_transaction_view(transaction_view, ts[-1])

    def autofill_transaction_view(self, transaction_view, transaction):
        transaction_view.replace_postings(transaction.postings)
        transaction_view.set_clearing(transaction.state)

    def update_transaction_view(self, unused_ignored=None):
        self.update_validation()
        k = self.transholder.get_data_for_transaction_record
        title, date, clear, statechar, lines = k()
        self.transaction_view.generate_record(
             title, date, clear, statechar, lines
         )

    def update_validation(self, grab_focus=False):
        try:
            self.transholder.validate(grab_focus=grab_focus)
            self.status.set_text("")
            self.add_button.set_sensitive(True)
            return True
        except common.TransactionInputValidationError as e:
            self.status.set_text(str(e))
            self.add_button.set_sensitive(False)
            return False

    def process_transaction(self):
        if not self.update_validation(True):
            return
        buf = self.transaction_view.get_buffer()
        text = buf.get_text(buf.get_start_iter(), buf.get_end_iter(), True)
        self.journal.add_text_to_file(text)
        self.reset_after_save()

    def reset_after_save(self):
        self.transholder.clear()
        self.transholder.title_grab_focus()
        self.status.set_text("Transaction saved")
        self.reload_completion_data()

    def save_preferences(self):
        if not self.successfully_loaded_accounts_and_commodities:
            return
        self.preferences["default_to_clearing"] = (
            self.transholder.clearing.get_state() !=
            self.transholder.clearing.STATE_UNCLEARED
        )
        if self.transholder.when.get_date() in (datetime.date.today(), None):
            del self.preferences["last_date"]
        else:
            self.preferences["last_date"] = (
                self.transholder.when.get_date()
            )
        self.preferences.persist()


def get_argparser():
    parser = argparse.ArgumentParser(
        'Add new transactions to your Ledger file',
        parents=[common_programs.get_common_argparser()]
    )
    parser.add_argument('--debug', dest='debug', action='store_true',
                        help='activate debugging')
    return parser


def main():
    args = get_argparser().parse_args()
    common.enable_debugging(args.debug)

    GObject.threads_init()

    journal, s = gui.load_journal_and_settings_for_gui(
        ledger_file=args.file,
        price_file=args.pricedb,
    )
    klass = AddTransApp
    win = klass(journal, s)
    win.connect("delete-event", Gtk.main_quit)
    GObject.idle_add(win.show_all)
    Gtk.main()
